irc/client_lookup_set.go

```diff
@@ -37,132 +36,108 @@ func ExpandUserHost(userhost string) (expanded string) {
 	return
 }
 
-// ClientLookupSet represents a way to store, search and lookup clients.
-type ClientLookupSet struct {
-	ByNickMutex sync.RWMutex
-	ByNick      map[string]*Client
+// ClientManager keeps track of clients by nick, enforcing uniqueness of casefolded nicks
+type ClientManager struct {
+	sync.RWMutex // tier 2
+	byNick      map[string]*Client
 }
 
-// NewClientLookupSet returns a new lookup set.
-func NewClientLookupSet() *ClientLookupSet {
-	return &ClientLookupSet{
-		ByNick: make(map[string]*Client),
+// NewClientManager returns a new ClientManager.
+func NewClientManager() *ClientManager {
+	return &ClientManager{
+		byNick: make(map[string]*Client),
 	}
 }
 
-// Count returns how many clients are in the lookup set.
-func (clients *ClientLookupSet) Count() int {
-	clients.ByNickMutex.RLock()
-	defer clients.ByNickMutex.RUnlock()
-	count := len(clients.ByNick)
+// Count returns how many clients are in the manager.
+func (clients *ClientManager) Count() int {
+	clients.RLock()
+	defer clients.RUnlock()
+	count := len(clients.byNick)
 	return count
 }
 
-// Has returns whether or not the given client exists.
-//TODO(dan): This seems like ripe ground for a race, if code does Has then Get, and assumes the Get will return a client.
-func (clients *ClientLookupSet) Has(nick string) bool {
+// Get retrieves a client from the manager, if they exist.
+func (clients *ClientManager) Get(nick string) *Client {
 	casefoldedName, err := CasefoldName(nick)
 	if err == nil {
-		return false
-	}
-	clients.ByNickMutex.RLock()
-	defer clients.ByNickMutex.RUnlock()
-	_, exists := clients.ByNick[casefoldedName]
-	return exists
-}
-
-// getNoMutex is used internally, for getting clients when no mutex is required (i.e. is already set).
-func (clients *ClientLookupSet) getNoMutex(nick string) *Client {
-	casefoldedName, err := CasefoldName(nick)
-	if err == nil {
-		cli := clients.ByNick[casefoldedName]
-		return cli
-	}
-	return nil
-}
-
-// Get retrieves a client from the set, if they exist.
-func (clients *ClientLookupSet) Get(nick string) *Client {
-	casefoldedName, err := CasefoldName(nick)
-	if err == nil {
-		clients.ByNickMutex.RLock()
-		defer clients.ByNickMutex.RUnlock()
-		cli := clients.ByNick[casefoldedName]
+		clients.RLock()
+		defer clients.RUnlock()
+		cli := clients.byNick[casefoldedName]
 		return cli
 	}
 	return nil
 }
 
-// Add adds a client to the lookup set.
-func (clients *ClientLookupSet) Add(client *Client, nick string) error {
-	nick, err := CasefoldName(nick)
-	if err != nil {
-		return err
-	}
-	clients.ByNickMutex.Lock()
-	defer clients.ByNickMutex.Unlock()
-	if clients.getNoMutex(nick) != nil {
-		return ErrNicknameInUse
+func (clients *ClientManager) removeInternal(client *Client) (removed bool) {
+	// requires holding ByNickMutex
```

Comment:
Mind updating the comment here? We can just say like `requires holding state mutex` or similar